### ISSUE TITLE
Sync API

### ISSUE THREAD

**Issue by LinusU** [danielgtaylor/aglio#111]
While not generally recommended there are some use cases which could really use a synchronous api. I think that we should add `.renderSync` which instead of a callback returns the html or throws an error.

This would be dependant on apiaryio/protagonist#69 which is merged but not yet published to npm.

**Comment by zdne** [comment]
> While not generally recommended ...

personally I am not really happy with introducing the sync() methods at all – as in protagonist I would love to hear some reasoning behind this...

**Comment by LinusU** [comment]
I think that the point of asynchronicity in Node.js is to delegate io to the kernel and that is really cool. libuv also provides methods for delegating work onto a background thread, with a default thread-pool of 4 workers.

**Protagonist** is doing 0% io and 100% cpu bound work and because of that it will need the cpu cycles, there is no going away from that. I guess that the decision to send it to a background thread was because it often took a noticeable amount of time and sending it to a background thread is a way to push that problem out of the way.

This article gives some insight into how node handles its threads: [On problems with threads in node.js](http://www.future-processing.pl/blog/on-problems-with-threads-in-node-js/)

I think it's important for protagonist to have a sync api because it isn't necessarily it's call to make in which thread to place the work. If I need to read a lot of markdown with protagonist I'm going to want to do my own dedicated threadpool or workerpool. To implement that I need a sync api.

**Jade** has a mix of both io and cpu bound work, but they have opted on only using `fs.*Sync` methods to make their api easier to use. The *seemingly asynchronous* api that aglio is using is actually very synchronous and the callback function is guaranteed to fire before the call to `render` is done.

e.g.

```javascript
jade.renderFile templatePath, locals, (err, html) ->
  console.log 'I will print first'
console.log 'I will print after'
```

**aglio** consist mostly of a composition of the two previously mentioned libraries and is thus in reality very synchronous. Reading all the template files, parsing them and then generating the html output is done **while blocking the main loop**. Reading all markdown includes is also down while blocking.

In essence the only thing happening asynchronous is parsing the markdown into an AST, what protagonist does.

So therefor, for the same reason as why I think protagonist should have a `*Sync` method, I think that aglio should have it as well.

**Comment by LinusU** [comment]
On a side note, I think that I would even more prefer that aglio accepted an AST and didn't depend on `protagonist` at all, but this isn't feasible since aglio is suppose to be out-of-the-box ready to go.

What I think would be cool would be to split aglio up into `aglio` and `aglio-renderer` (or similar). `aglio-renderer` would be responsible for transforming an AST into HTML, and nothing else. `aglio` would hold all the "glue".

**Comment by danielgtaylor** [comment]
@LinusU I'm not opposed to adding a sync API if it's really useful and the upstream libs add one that I can piggyback off of. It also makes testing a little easier for downstream projects.

The point of using threads for CPU-bound tasks is that it will run in a background thread while the main thread will still process events. For a server-side application this is pretty important, because it means you can offload reading files/streams, parsing them, rendering, etc all while still handling incoming requests and e.g. adding them to a queue. I'm a bit surprised at how Jade is implemented to be honest, but other theme engines could handle this in a better way.

You should take a look at the new [theme-engine](https://github.com/danielgtaylor/aglio/tree/theme-engine) and [olio-theme](https://github.com/danielgtaylor/aglio/tree/olio-theme) branches. They split things exactly as you suggest, and make it possible to create new themes which can render given an AST (and in the future probably serialized [refract](https://github.com/refractproject/refract-spec) elements). That includes things like rendering in the browser potentially. You can play with it by doing `sudo npm install -g aglio@beta`. Rendering is also 10x faster in that branch with a cold cache, and more like 100x faster with a hot cache.

**Comment by LinusU** [comment]
@danielgtaylor Sounds good, I can send a pull request as soon as they have published a new release on npm.

I think the problem is that if you render more than 3 blueprints with aglio at the same time it will lock up everything else until it's done, which is something that I think most user aren't aware of. Maybe this isn't such a huge problem, but for users who do know how threads are used in Node, I think it's wrong to assume that they want to do the parsing on a background thread. Basically what I'm saying it that in this case, choice is good.

Exciting work in the branches, although the theme engine is still mixing sync and async functions. My recommendation would be to compile the less into css at build time, instead of when the template is about to be rendered. `less` would then be a `devDependency` instead, saving the install for the clients. This is similar to how you compile the coffeescript to javascript before publishing, instead of doing it at runtime.

Since `less` is the only asynchronous step, the whole thing could then be synchronous.

**Comment by LinusU** [comment]
@danielgtaylor I see that you switched to `drafter.js` instead of `protagonist` in the beta. There is no reason for drafter.js to have a asynchronous api.

See apiaryio/boutique.js#57

**Comment by zdne** [comment]
@LinusU @danielgtaylor for the record: Drafter.js is a temporary solution before we implement Boutique.js functionality in "C++" Drafter. 

After that Protagonist will be once again the Node.js binding for the API Blueprint parser (harness). 

So please treat everything related to Boutique.js / Drafter.js as an interim solution.

**Comment by LinusU** [comment]
Protagonist has published a new version `0.20.0` which contains `renderSync`. I've opened #120 which adds it to aglio as well :)
